Index: fieldworks-enchant-1.6.1/src/pwl.c
===================================================================
--- enchant-1.6.1/src/pwl.c
+++ fieldworks-enchant-1.6.1/src/pwl.c
@@ -284,7 +284,7 @@
	char* line;
	size_t line_number = 1;
	FILE *f;
-	struct stat stats;
+	GStatBuf stats;

	if(!pwl->filename)
		return;
@@ -397,7 +397,7 @@
		f = enchant_fopen(pwl->filename, "a");
		if (f)
			{
-				struct stat stats;
+				GStatBuf stats;

				enchant_lock_file (f);
				if(g_stat(pwl->filename, &stats)==0)
@@ -440,7 +440,7 @@
					const gunichar BOM = 0xfeff;
					char * filestart, *searchstart, *needle;
					char * key;
-					struct stat stats;
+					GStatBuf stats;

					enchant_lock_file (f);
					key = g_strndup(word, len);
